### ISSUE TITLE
since micronaut 2.3.4: `IllegalStateException: No delegate constructor for [simple type, class my.org.MyReqBody]`

### ISSUE THREAD

**Issue by rdesgroppes** [micronaut-projects/micronaut-core#5088]
_A workaround exists (see somewhere below)._

### Steps to Reproduce

1. have a project for a service powered by micronaut 2.3.3
2. with a request handler like:
```java
  public Single<MyRespBody> myRequestHandler(
      @RequestBody(
              content = @Content(schema = @Schema(implementation = MyReqBody.class)),
              required = true)
          @Valid
          final MyReqBody body) {
    return Single.just(doSthWith(body));
  }
```
3. and a data class like:
```java
@Introspected
@Schema(description="...", example="...")
public final class MyReqBody {
  private final List<MyItem> items;

  public MyReqBody(final List<MyItem> items) {
    this.items = items;
  }

  @ArraySchema(arraySchema = @Schema(description = "..."))
  @Size(min = 2, max = 10000)
  @NotEmpty
  @Valid
  public List<MyItem> getItems() {
    return items;
  }
}
```
4. service works as expected: `MyReqBody` objects are created from a list of `MyItem`s.
5. upgrade to micronaut 2.3.4 (or 2.4.0)

### Expected Behaviour

`MyReqBody` objects are created from a list of `MyItem`s as with earlier versions of micronaut.

### Actual Behaviour

```
[ERROR] [2021-...] [default-nioEventLoopGroup-...-...] [RoutingInBoundHandler] [correlation-id=my-client-cid] Unexpected error occurred: No delegate constructor for [simple type, class my.org.MyReqBody]
    java.lang.IllegalStateException: No delegate constructor for [simple type, class my.org.MyReqBody]
    	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator._createUsingDelegate(StdValueInstantiator.java:530)
    	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingDelegate(StdValueInstantiator.java:298)
    	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1314)
    	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:331)
    	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:164)
    	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4495)
    	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2728)
    	at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:3181)
    	at io.micronaut.jackson.convert.JacksonConverterRegistrar.lambda$jsonNodeToObjectConverter$4(JacksonConverterRegistrar.java:229)
    	at io.micronaut.core.convert.DefaultConversionService.convert(DefaultConversionService.java:135)
    	at io.micronaut.context.env.DefaultEnvironment.convert(DefaultEnvironment.java:289)
    	at io.micronaut.core.convert.ConversionService.convert(ConversionService.java:93)
    	at io.micronaut.http.server.netty.NettyHttpRequest.lambda$null$2(NettyHttpRequest.java:265)
    	at java.util.HashMap.computeIfAbsent(Unknown Source)
    	at io.micronaut.http.server.netty.NettyHttpRequest.lambda$getBody$3(NettyHttpRequest.java:265)
    	at java.util.Optional.flatMap(Unknown Source)
    	at io.micronaut.http.server.netty.NettyHttpRequest.getBody(NettyHttpRequest.java:265)
    	at io.micronaut.http.bind.binders.ParameterAnnotationBinder.lambda$bind$0(ParameterAnnotationBinder.java:93)
    	at io.micronaut.core.bind.ArgumentBinder$BindingResult.isPresentAndSatisfied(ArgumentBinder.java:105)
    	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:277)
    	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
    	at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$12(RoutingInBoundHandler.java:1394)
    	at io.reactivex.internal.operators.flowable.FlowableDefer.subscribeActual(FlowableDefer.java:35)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.micronaut.reactive.rxjava2.RxInstrumentedFlowable.subscribeActual(RxInstrumentedFlowable.java:57)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.reactivex.internal.operators.flowable.FlowableDefer.subscribeActual(FlowableDefer.java:42)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.micronaut.reactive.rxjava2.RxInstrumentedFlowable.subscribeActual(RxInstrumentedFlowable.java:57)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14868)
    	at io.micronaut.configuration.metrics.binder.web.WebMetricsPublisher.subscribe(WebMetricsPublisher.java:155)
    	at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.micronaut.reactive.rxjava2.RxInstrumentedFlowable.subscribeActual(RxInstrumentedFlowable.java:57)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:50)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.micronaut.http.server.context.ServerRequestContextFilter.lambda$doFilter$0(ServerRequestContextFilter.java:62)
    	at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
    	at io.micronaut.reactive.rxjava2.RxInstrumentedFlowable.subscribeActual(RxInstrumentedFlowable.java:57)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
    	at io.reactivex.Flowable.subscribe(Flowable.java:14868)
    	at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildExecutableRoute$8(RoutingInBoundHandler.java:1099)
    	at io.micronaut.web.router.DefaultUriRouteMatch$1.execute(DefaultUriRouteMatch.java:80)
    	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:118)
    	at io.micronaut.http.server.netty.RoutingInBoundHandler$2.doOnComplete(RoutingInBoundHandler.java:1025)
    	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
    	at io.micronaut.http.server.netty.jackson.JsonContentProcessor$1.doOnComplete(JsonContentProcessor.java:140)
    	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
    	at java.util.Optional.ifPresent(Unknown Source)
    	at io.micronaut.core.async.processor.SingleThreadedBufferingProcessor.doOnComplete(SingleThreadedBufferingProcessor.java:48)
    	at io.micronaut.jackson.parser.JacksonProcessor.doOnComplete(JacksonProcessor.java:137)
    	at io.micronaut.core.async.subscriber.SingleThreadedBufferingSubscriber.onComplete(SingleThreadedBufferingSubscriber.java:70)
    	at io.micronaut.http.server.netty.jackson.JsonContentProcessor.doOnComplete(JsonContentProcessor.java:165)
    	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
    	at io.micronaut.http.netty.reactive.HandlerPublisher.complete(HandlerPublisher.java:416)
    	at io.micronaut.http.netty.reactive.HandlerPublisher.handlerRemoved(HandlerPublisher.java:403)
    	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)
    	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
    	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
    	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
    	at io.micronaut.http.netty.stream.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:429)
    	at io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:294)
    	at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:257)
    	at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:121)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
    	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
    	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
    	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
    	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
    	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
    	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
    	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
    	at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
    	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
    	at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:200)
    	at io.netty.handler.flow.FlowControlHandler.read(FlowControlHandler.java:139)
    	at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
    	at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
    	at io.micronaut.http.netty.reactive.HandlerPublisher.requestDemand(HandlerPublisher.java:163)
    	at io.micronaut.http.netty.stream.HttpStreamsHandler$2.requestDemand(HttpStreamsHandler.java:248)
    	at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.receivedDemand(HandlerPublisher.java:547)
    	at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.lambda$request$0(HandlerPublisher.java:474)
    	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
    	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
    	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
    	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
    	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
    	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    	at java.lang.Thread.run(Unknown Source)
```

### Workaround

Replacing above mentioned `List<MyItem>` by `MyItem[]` works around the issue in my case.

### Environment Information

- **Operating System**: Any (Ubuntu, Alpine, etc.)
- **Micronaut Version:** OK up to 1.3.3; breaking change observed with both 1.3.4 and 1.4.0
- **JDK Version:** OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~20.04-b08)

**Comment by jameskleeh** [comment]
@rdesgroppes Is `MyItem` annotated with `@Introspected`? Can you provide an example? I can't reproduce

**Comment by jameskleeh** [comment]
I added tests for this scenario. If you can tell me what needs to change to make them fail I'll take another look

**Comment by rdesgroppes** [comment]
> @rdesgroppes Is `MyItem` annotated with `@Introspected`? Can you provide an example? I can't reproduce

@jameskleeh `MyItem` looks like:
```java
@Introspected
@Schema(description = "...")
public final class MyItem {
  private final Double aDouble;
  private final Double anotherDouble;

  public MyItem(final Double aDouble, final Double anotherDouble) {
    this.aDouble = aDouble;
    this.anotherDouble = anotherDouble;
  }

  @Schema(description = "...", example = "...")
  @DecimalMin(value = "-42.0")
  @DecimalMax(value = "42.0")
  @NotNull
  public Double getADouble() {
    return aDouble;
  }

  @Schema(description = "...", example = "...")
  @DecimalMin(value = "-666.0")
  @DecimalMax(value = "666.0")
  @NotNull
  public Double getAnotherDouble() {
    return anotherDouble;
  }
}
```

**Comment by goeh** [comment]
I saw the same exception in my KafkaListener after I upgraded to 2.3.4.
The message class has a private default constructor. If I change it to public, the exception disappears.
I thought the problem was on my side, that the constructor needs to be protected or public. So I just changed it to public and the problem went away. I didn't think of reporting it at the time.
```
@Introspected
public class GenericCrmEvent extends AbstractCrmEvent<Map<String, Object>> {

    private GenericCrmEvent() {}

    public GenericCrmEvent(Map<String, Object> data) {
        super(data);
    }

    public GenericCrmEvent(Map<String, Object> data, Map<String, String> attributes) {
        super(data, attributes);
    }
}
```

**Comment by fwpascual** [comment]
I am experiencing similar issues (not an issue on v2.3.3, but is an issue with v2.3.4 and v2.4.0).  A simple example Controller (in Kotlin) that demonstrates the issue:

Create a project using:
`mn create-app --build=gradle_kotlin --jdk=11 --lang=kotlin --test=junit com.example.demo`

Add a Controller:
```
@Validated
@Controller("example/v1")
class ExampleController {
    @Post("/good")
    fun postExampleGood(@Valid @Body body: Good) {
        println("Body received: $body")
    }

    @Post("/bad")
    fun postExampleBad(@Valid @Body body: Bad) {
        println("Body received: $body")
    }
}

@Introspected
data class Good(@field:Valid val sample: Sample)

@Introspected
data class Bad(@field:Valid val samples: List<Sample>)

@Introspected
data class Sample(@field:Positive val x: Int, @field:Positive val y: Int)
```

Call with curl:
- Good:
```
curl http://localhost:8080/example/v1/good -i --request POST --header 'Content-Type: application/json' --data-raw '{ "sample": { "x": 1, "y": 2 } }'
```
- Bad
```
curl http://localhost:8080/example/v1/bad -i --request POST --header 'Content-Type: application/json' --data-raw '{ "samples": [{ "x": 1, "y": 2 }] }'
```

On v2.3.4 and v2.4.0, the `/example/v1/bad` path returns:
```
{"message":"Internal Server Error: No delegate constructor for [simple type, class com.example.Bad]","_links":{"self":{"href":"/example/v1/bad","templated":false}}}
```

**Comment by jameskleeh** [comment]
This was fixed for 4.1, however note for `MyReqBody` the constructor needs annotated with `@JsonCreator`
